### ISSUE TITLE
make_csv_dataset with num_epochs=1,2,... (finite) produces (None,) shape tensors

### ISSUE THREAD

**Issue by srfrnk** [tensorflow/tensorflow#33079]
The following CoLab and Github Gist show the issue:
[CoLab](https://colab.research.google.com/gist/srfrnk/5744026f803b57254b5ba6162dd16bf8/csv.ipynb)
[Github Gist](https://gist.github.com/srfrnk/5744026f803b57254b5ba6162dd16bf8#file-csv-ipynb)

It is expected that the same shape will be returned regardless of num_epochs.
Currently if num_epochs=1 the returned tensors all have shape (None,)
Only when num_epochs=None meaning infinite repeat will the correct shape be returned.

**Comment by CarMiranda** [comment]
From the docstring:
> A dataset, where each element is a (features, labels) tuple that corresponds
    to a batch of `batch_size` CSV rows.

So I guess you would expect to have elements with shape `batch_size`?

-- The following is explained more concisely (and precisely for your case) [here](https://github.com/tensorflow/tensorflow/blob/4aa790771fef28b68cb7703e72a0824e12c3732a/tensorflow/python/data/experimental/ops/readers.py#L547) in the source code. Shape inference has lots to do with buffered reading and prefetching, in order to achieve good performance, but I won't get into details. You can find more [here](https://www.tensorflow.org/guide/data_performance), for example. --

On the one hand, when `num_epochs` is `None`, the dataset loops indefinitely, so shape inference can be done at this step: one can be certain there will not be any incomplete batch since there is a potentially infinite amount of data rows. Thus, every batch will contain exactly `batch_size` elements (static, constant shape).
On the other hand, when `num_epochs` is an integer, one cannot be sure every batch will have exactly `batch_size` elements (e.g. the last batch in your epoch), so the shape is dynamic, and hence the use of `None`.

Moreover, when using the elements from, for example, the `TakeDataset` you have created in your example, shape is dynamically inferred and shown:

```python
>>> for el in ds1: print(el)
OrderedDict([('PassengerId', <tf.Tensor: id=2278, shape=(1,), ...])
```

Could you explain why this would be an issue/bad practice in a common use case?

**Comment by srfrnk** [comment]
@CarMiranda Thanks for your  input.
Indeed I would expect the tensors to be of shape (`batch_size`,..)
The `batch_size` should be already defined he same way for when `num_epochs` is None or a finite number am I wrong?
If so - the tensors returned should at least have the `batch_size` as a "known" dimension. You might argue about the other dimensions for sure but not the first one. Am I wrong?

In my case I was using the tensor shape to create a zeros tensor with the same shape.
This code worked for the train set - read with repeat - but failed for the prediction set which was read without repeat.

I did solve my issue by using the tensorflow.shape(...) function which returned a dynamic shape and I passed that into tensorflow.zeros(...) instead. 

I still think this is a bug that might hinder someone in the future.

**Comment by CarMiranda** [comment]
> If so - the tensors returned should at least have the `batch_size` as a "known" dimension. You might argue about the other dimensions for sure but not the first one. Am I wrong?

Actually, in most cases, I personally would expect all of the other dimensions to be known, but not the batch size... For example, when loading the MNIST data set from `tensorflow_datasets` and using `batch(N)`, the batch dimension is `None`, not `N`. In fact, columns/variables/features/image dimensions are something that remain constant over time, while the number of rows in a data set does not.

The method `make_csv_dataset` behaves like `tf.data.Dataset.batch(batch_size, drop_remainder=False)` when `num_epochs` is set (so it uses dynamic shaping) and like `tf.data.Dataset.bacth(batch_size, drop_remainder=True)` when `num_epoch=None`, which is the expected/standard behaviour. You can see [this](https://github.com/tensorflow/tensorflow/issues/18226) issue (#18226) which led to the `drop_remainder` option.

Some more practical considerations: Say my data set contains 10 columns and 10 000 rows. I set batch_size=64 because that's a nice value in my case (e.g. taking hardware into account). After 156 batches, my model has seen 9 984 rows, training on fully-sized batches of 64 rows. But right at the end, the last batch has only 16 rows. What do we fill the rest of the tensor with? Zeros? Ones? Random rows? Or should we just drop these from the training? (which is an option in the `tf.data.Dataset.batch(batch_size, drop_remainder)` method, from which `make_csv_dataset` makes use). Well, with dynamic batch size, that is no problem, we just set `batch_size=16` when needed.

But hey, I could set `batch_size=50`, or any other divisor of 10 000 and the problem is gone. However, I would be bounding the model training to a specific (version of the) data set. If the data set evolves and adds 1234 rows, then I would have to change the `batch_size`. I could even get more rows in the dataset _while_ training, between epochs: in production, number of rows can vary as data comes in.

---

> In my case I was using the tensor shape to create a zeros tensor with the same shape.

Is there any reason not to use `new_tensor = tf.zeros_like(tensor_to_copy)`?

**Comment by srfrnk** [comment]
So basically you mean that tensor shapes should be undetermined with various . I would think it should be mentioned in the docs at least?

**Comment by jvishnuvardhan** [comment]
@srfrnk Is this still an issue for you? If this was already resolved for you then please close the issue. Thanks!
